Schema for bug files

wasowski opened the pull request:
I made a simple json schema for our YAML files and a script to validate them.  Please review and merge.  (the last two sections are left free, because I am not yet familiar with them sufficiently; we can do them in a separate PR).

Probably, one reviewer is enough, so I assigned a bit opportunistically :)

git-afsantos commented:
I have not tried the script, although it seems simple enough. I have left a few comments on the schema.

Perhaps by using a `description` for each `property`, we could eliminate the need for `taxonomy-description.txt` altogether (i.e., make the schema self-documenting). Or would you rather have a separate documentation file?

ChrisTimperley commented on scripts/robust.json:
```diff
@@ -0,0 +1,74 @@
+    "system"         : { "type" : "string", "maxLength" : 40 },
+    "severity"       : { "enum": [ "error", "warning", "convention-violation", "bad-smell", "minor-issue" ] },
+    "links"          : { "type" : "array", "items" : { "type" : "string", "format" : "uri" } },
+    "bug"            : { "type" : "object",
```
Minor nitpick: I'd try to avoid aligning the `:` characters in this way. If a field in `properties` is added or changed, then you would potentially need to change the whitespacing for every line covered by `properties`, which leads to bigger diffs that are a little harder to parse.

ChrisTimperley commented on scripts/validate.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
Can we integrate this script into the CI?

gavanderhoorn replied:
I would say we must integrate this with the current CI setup.

Anything else would not make sense to me.

wasowski commented:
Hold your horses 🐎! You're gonna get a million of errors and stop merging
PRs (or start ignoring CI). Let's see for a few days how many things are
flagging errors. Also let me fix the script slightly to print the name of a
broken bug file (otherwise we don't know which file is broken; so far
validate.sh just says that /dev/stdin is broken :) ).

On Wed, May 15, 2019, 20:55 G.A. vd. Hoorn <notifications@github.com> wrote:

> *@gavanderhoorn* commented on this pull request.
> ------------------------------
>
> In scripts/validate.sh
> <https://github.com/robust-rosin/robust/pull/288#discussion_r284403101>:
>
> > @@ -0,0 +1,11 @@
> +#!/bin/bash
>
> I would say we must integrate this with the current CI setup.
>
> Anything else would not make sense to me.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/robust-rosin/robust/pull/288?email_source=notifications&email_token=AANMEHERGCW6EZ2GQST67S3PVRMCZA5CNFSM4HNFJ3K2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBYX34JA#discussion_r284403101>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AANMEHATGC5HE4ULFY7F57DPVRMCZANCNFSM4HNFJ3KQ>
> .
>

wasowski commented:
I have fixed the white space issues, other formatting issues (line length, etc.). I also improved the reporting from validate.sh (so that you see a flie name that is wrong). 

I am not sure if the script returns a meaningful error code for Travis (but any work of this kind belongs, IMHO, to a separate PR, regarding travis integration). 

Re. taxonomy-description: sounds like a good idea, but a major work, that I am not so willing to undertake now.  Let's do it later (#290).

IMHO, to merge this, we should still agree on #289 with Chris, regarding nulls vs empty strings.

gavanderhoorn commented:
> IMHO, to merge this, we should still agree on #289 with Chris, regarding nulls vs empty strings.

`null`s make sense to me as that is a clear indication that something is not present on purpose.

leaving a field empty could mean two things:

 1. something is unknown -> on purpose
 1. someone forgot to populate the field -> mistake

`null` avoid that ambiguity.

gavanderhoorn commented on scripts/robust.json:
```diff
@@ -0,0 +1,156 @@
+        "license" : { 
+          "type" : "array", 
+          "items" : { 
+            "enum" : [ "BSD", "GPLv3", "LGPLv3" ] 
```
```suggestion
            "enum" : [ "BSD", "GPLv3", "LGPLv3", "Apache2.0" ] 
```

gavanderhoorn commented on scripts/validate.sh:
```diff
@@ -0,0 +1,11 @@
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+SCHEMA="${DIR}/robust.json"
+echo Checking schema compliance for $1
+any-json --input-format=yaml $1 | ajv --verbose --errors=js -s ${SCHEMA} -d /dev/stdin | sed "s/\/dev\/stdin/File/"
```
```suggestion
any-json --input-format=yaml $1 | ajv --verbose --errors=text -s ${SCHEMA} -d /dev/stdin | sed "s/\/dev\/stdin/File/"
```

We're humans and we don't have an automated reporting solution in place to parse the JS output for us.

`text` gives us:

> human readable error messages with data paths

wasowski replied:
I thought you would prefer structured messages (because I thought that maybe travis knows what to do with them). If not, I prefer human readable too :)

gavanderhoorn commented:
> I am not sure if the script returns a meaningful error code for Travis (but any work of this kind belongs, IMHO, to a separate PR, regarding travis integration).

testing Travis integration I notice that the pass through `any-json` is not conducive to performance: the manual for-loop with `xargs` and feeding `ajv-cli` one file at a time is slow.

Regular CI runs in about 30 secs. With all files checked by `ajv-cli` we run into 3 mins, easy. See [robust-rosin/robust/runs/129619064](https://github.com/robust-rosin/robust/runs/129619064) for an example.

Output also doesn't mention line nrs or filenames, so it's difficult to interprete the output correctly.

Was just a quick test though, things can probably be improved.

gavanderhoorn commented:
@wasowski: my :horse:s are waiting :)

What is the status of your PR?

wasowski commented:
I think I have addressed all the functional requests.  Here are the main points:

1. **Moved the descriptions into the "description" field of the schema (and deleted the old text file).** I did not make any substantial changes to the descriptions.  I tried to keep them in sync with schema.  At times they do sound like they should be updated slightly (along with the schema).  I guess we will do this, when we start to enforce the schema, and see that it requires things that are unreasonable in concrete situations. 

2. **Allowed nulls when the format explicitly allows them, disallow empty strings when they are not allowed**. Unfortunately, it appears that an empty entry is equivalent to null (not to an empty string). So I was unable to forbid empty strings for fields, where null is allowed.  This will make it hard to chase empty entries (but a grep might just do this).

3. **Fixed any bugs I have identified while playing with this.**

4. **Updated the test example mavros/08* to validate** (mostly brough the things from the other PR that I accidentally duplicated here)

5. **Switched to textual messages**.   I could not figure out any way to force ajv to report line numbers (this does suck, but field names identify places in yaml files pretty well -- the problem is that validation happens via translation to json, so ajv has no access to original YAML line numbers, I believe).

Speed issues? I think 3 minutes is acceptable for as at this point?  I see no easy way to fix it.

I believe we could merge, and start checking our files.

wasowski commented:
Do we have an OK to merge?

gavanderhoorn commented:
> Can we integrate this with the Travis config?

We can, but the current implementation is actually quite slow.

> Do we have an OK to merge?

@wasowski: you may want to update the author of the commits in this PR: the latter ones appear to have been made with your gmail address as the `AUTHOR_EMAIL`.
